packages/scm/src/browser/scm-service.ts

```diff
@@ -0,0 +1,325 @@
+    private onDidChangeSelectedRepositoriesEmitter = new Emitter<ScmRepository>();
+    private onDidAddProviderEmitter = new Emitter<ScmRepository>();
+    private onDidRemoveProviderEmitter = new Emitter<ScmRepository>();
+
+    readonly onDidChangeSelectedRepositories: Event<ScmRepository> = this.onDidChangeSelectedRepositoriesEmitter.event;
+
+    constructor() {
+        this.disposableCollection.push(this.onDidChangeSelectedRepositoriesEmitter);
+        this.disposableCollection.push(this.onDidAddProviderEmitter);
+        this.disposableCollection.push(this.onDidRemoveProviderEmitter);
+    }
+
+    get repositories(): ScmRepository[] {
+        return [...this._repositories];
+    }
+
+    get selectedRepository(): ScmRepository | undefined {
+        return this._selectedRepository;
+    }
+
+    set selectedRepository(repository: ScmRepository | undefined) {
+        this._selectedRepository = repository;
+        if (repository) {
+            this.onDidChangeSelectedRepositoriesEmitter.fire(repository);
+        }
+    }
+
+    get onDidAddRepository(): Event<ScmRepository> {
+        return this.onDidAddProviderEmitter.event;
+    }
+
+    get onDidRemoveRepository(): Event<ScmRepository> {
+        return this.onDidRemoveProviderEmitter.event;
+    }
+
+    registerScmProvider(provider: ScmProvider): ScmRepository {
+
+        if (this.providerIds.has(provider.id)) {
+            throw new Error(`SCM Provider ${provider.id} already exists.`);
+        }
+
+        this.providerIds.add(provider.id);
+
+        const disposable: Disposable = Disposable.create(() => {
+            const index = this._repositories.indexOf(repository);
+            if (index < 0) {
+                return;
+            }
+            this.providerIds.delete(provider.id);
+            this._repositories.splice(index, 1);
+            this.onDidRemoveProviderEmitter.fire(repository);
+        });
+
+        const repository = new ScmRepositoryImpl(provider, disposable);
+
+        this._repositories.push(repository);
+        this.onDidAddProviderEmitter.fire(repository);
+
+        // automatically select the first repository
+        if (this._repositories.length === 1) {
+            this.selectedRepository = repository;
+        }
+
+        return repository;
+    }
+
+    dispose(): void {
+        this.disposableCollection.dispose();
+    }
+}
+
+class ScmRepositoryImpl implements ScmRepository {
+
+    private _onDidFocus = new Emitter<void>();
+    readonly onDidFocus: Event<void> = this._onDidFocus.event;
+
+    private _selected = false;
+    get selected(): boolean {
+        return this._selected;
+    }
+
+    private _onDidChangeSelection = new Emitter<boolean>();
+    readonly onDidChangeSelection: Event<boolean> = this._onDidChangeSelection.event;
+
+    readonly input: ScmInput = new ScmInputImpl();
+
+    constructor(
+        public readonly provider: ScmProvider,
+        private disposable: Disposable
```

Comment:
don't forget to collect all emitters in disposable, please review code for all such places, it leaks listeners

Reply:
done